File: recipes/help.py
from __future__ import annotations

from utils.loader import GobiFile
from utils.recipes import GobiError, Action, Recipe, load_recipe

def surround_print(text: str) -> None:
    str_len = len(text)
    print("#" * (str_len + 4))
    print(f"# {text} #")
    print("#" * (str_len + 4))

class HelpAction(Action):
    def __init__(self) -> None:
        super().__init__()
        self.name = "help"
        self.subname = "help"

    def help(self) -> str:
        return "How clever of you to find this help menu!"

    def help_menu(self) -> None:
        print("""
Usage: gobi <project list...>? help [recipe|action] <recipe name|action name...>?

default: prints this help menu

recipe: prints the help menu for all recipes loaded in the current project
        <recipe name...>: prints the help menu for the specified recipes

action: prints the help menu for all actions loaded in the current project
        <action name...>: prints the help menu for the specified actions
""")

    def run(
        self,
        gobi_file: GobiFile,
        recipes: dict[str, Recipe],
        actions: list[Action],
        args: list[str],
    ) -> GobiError | None:
        if args == []:
            self.help_menu()
        elif args[0] == "recipe":
            if len(args) == 1:
                print("Help for all loaded recipes:\n")
                for recipe in sorted(recipes.keys()):
                    surround_print("Help menu for: " + recipe)
                    print(f"\n{recipes[recipe].help().strip()}\n")
            else:
                for recipe in args[1:]:
                    surround_print("Help menu for: " + recipe)
                    if recipe not in recipes:
                        recipe_obj = load_recipe(recipe)
                        if recipe_obj is not None:
                            print(f"Unknown recipe: {recipe}")
                    else:
                        print(f"\n{recipes[recipe].help().strip()}\n")
        elif args[0] == "action":
            if len(args) == 1:
                print("Help for all loaded actions:")
                for action in sorted(actions, key=lambda a: a.subname + a.name):
                    surround_print("Help menu for: " + action.name)
                    print(f"\n{action.help().strip()}\n")
            else:
                for action in args[1:]:
                    possible_actions = list(filter(lambda a: a.subname == action, actions))
                    if len(possible_actions) == 0:
                        return GobiError(self, 1, f"Unknown action: {action}")
                    for action in possible_actions:
                        surround_print("Help menu for: " + action.name)
                        print(f"\n{action.help().strip()}\n")

class HelpRecipe(Recipe):
    def __init__(self):
        self.name = "help"

    def help(self) -> str:
        return "Generates the help action for gobi."

    def create_actions(self, gobi_file: GobiFile) -> GobiError | list[Action]:
        return [HelpAction()]


def create() -> Recipe:
    return HelpRecipe()

File: lib/recipes/src/help.rs
use gobi_lib::{file::GobiFile, recipes::*, GobiError, GobiResult};

pub struct HelpAction {
    recipe_manager: SharedRecipeManager,
}

impl HelpAction {
    fn new(recipe_manager: &SharedRecipeManager) -> Self {
        Self {
            recipe_manager: recipe_manager.clone(),
        }
    }
}

impl IAction for HelpAction {
    fn get_name(&self) -> &str {
        "help"
    }

    fn get_subname(&self) -> &str {
        "help"
    }

    fn get_help(&self) -> &str {
        "
Usage: gobi <project list...>? help [mode]? [<action_name>+]?

mode can be 'recipe' or 'action', if ommited will default to 'action'

With no arguments, will print the help menu for all actions/recipes, otherwise only prints the help for the specified action/recipe
"
    }

    fn run(&self, actions: &Vec<Action>, args: Vec<String>) -> GobiResult<()> {
        enum Mode {
            Action,
            Recipe,
        }

        let mut args = args.iter().peekable();

        let mode = match args.peek() {
            None => Mode::Action,
            Some(mode) => match mode.as_str() {
                "recipe" => {
                    args.next();
                    Mode::Recipe
                }
                "action" => {
                    args.next();
                    Mode::Action
                }
                _ => Mode::Action,
            },
        };

        let names: &mut dyn Iterator<Item = String> = if args.len() == 0 {
            &mut actions.iter().map(|a| a.get_name().into())
        } else {
            &mut args.map(|a| a.clone())
        };
        let messages: &mut dyn Iterator<Item = GobiResult<String>> = match mode {
            Mode::Recipe => &mut names.map(|rname| -> GobiResult<String> {
                let recipe = match self.recipe_manager.get_recipe(&rname) {
                    None => {
                        return Err(GobiError {
                            code: 1,
                            msg: format!("Recipe '{}' not found", rname),
                        })
                    }
                    Some(r) => r,
                };
                Ok(format!(
                    "{}\n# {} #\n{}\n{}",
                    "#".repeat(recipe.get_name().len() + 4),
                    recipe.get_name(),
                    "#".repeat(recipe.get_name().len() + 4),
                    recipe.get_help().trim()
                ))
            }),
            Mode::Action => &mut names.map(|aname| -> GobiResult<String> {
                let action = find_action(&aname, actions)?;
                let name = action.get_name();
                Ok(format!(
                    "{}\n# {} #\n{}\n{}",
                    "#".repeat(name.len() + 4),
                    action.get_name(),
                    "#".repeat(name.len() + 4),
                    action.get_help().trim()
                ))
            }),
        };

        let (err, help) = messages.fold(
            (
                GobiError {
                    code: 0,
                    msg: String::new(),
                },
                String::new(),
            ),
            |(acce, acch), m| match m {
                Err(e) => (acce.merge(e), acch),
                Ok(h) => (acce, format!("{}\n\n{}", acch, h)),
            },
        );

        if err.code != 0 {
            return Err(err.clone());
        }

        println!("{}", help.trim());
        Ok(())
    }
}

struct HelpRecipe {}

impl HelpRecipe {
    fn new() -> HelpRecipe {
        HelpRecipe {}
    }
}

impl IRecipe for HelpRecipe {
    fn get_name(&self) -> &str {
        "help"
    }

    fn get_help(&self) -> &str {
        "Generates the help action for gobi."
    }

    fn create_actions(
        &self,
        recipe_manager: &SharedRecipeManager,
        _gobi_file: &GobiFile,
    ) -> GobiResult<Vec<Action>> {
        return Ok(vec![ActionWrapper::new(HelpAction::new(recipe_manager))]);
    }
}

pub fn register_recipes(recipe_manager: &mut RecipeManager) {
    recipe_manager.register_recipe(RecipeWrapper::new(HelpRecipe::new()));
}
